core/src/main/scala/spinal/core/sim/SpinalSpec.scala

```diff
@@ -0,0 +1,276 @@
+    ref match {
+      case nameable: Nameable => nameable.setName(name)
+      case _                  =>
+    }
+    ref
+  }
+
+  /** Override it to define the default seed for tests (else seed seed is not fixed)
+    *
+    * It can be overriden from sbt command: `testOnly Myspec -- -Dseed=11`.
+    */
+  val defaultSeed: Option[Int] = None
+
+  // Private else user could modify it in a test, leading to strange behavior.
+  private var seed: Option[Int] = None
+
+  /** Seed can be set with `testOnly MyComponentSpec -- -Dseed=123456789` */
+  final override def beforeAll(configMap: ConfigMap): Unit = {
+    seed = configMap
+      .get("seed")
+      .map(_.asInstanceOf[String].toInt)
+      .orElse(defaultSeed)
+  }
+
+  /** Override it to define the default SpinalSimConfig (else defaults to SimConfig) */
+  def defaultConfig: SpinalSimConfig = SimConfig
+
+  /** Override it to false to disable caching of bench compilation */
+  val defaultCaching: Boolean = true
+
+  /** Define a group of tests that can be run on several benches */
+  case class TestGroup[SoftDut](tests: Bench[SoftDut] => Unit) extends Nameable {
+
+    /** Run the tests on (a) named (in a single `val`) benche(s) */
+    def runWith(benches: Bench[SoftDut]*): Unit = {
+      getName - {
+        for (bench <- benches) {
+          bench.getName - {
+            tests(bench)
+          }
+        }
+      }
+    }
+
+    /** Run the tests on a sequence of benches */
+    def runWithAll(namedBenches: Seq[(Bench[SoftDut], String)]): Unit = {
+      getName - {
+        for ((bench, name) <- namedBenches) {
+          name - {
+            tests(bench)
+          }
+        }
+      }
+    }
+  }
+
+  /** Define a group of tests that can be run with several component configurations
+    *
+    * {{{
+    * val myTests =
+    *  TestGroupCfg { (cfg: MyComponentConfig) =>
+    *    bench(new MyComponent(cfg))
+    *  } { (cfg, it) =>
+    *    it should "do nothing" in { dut => }
+    *  }
+    * }}}
+    *
+    * @param benchBuilder the function to create a bench from the component configuration
+    * @param tests tests to run, which can be conditional over the component configuration
+    */
+  case class TestGroupCfg[Cfg, SoftDut](benchBuilder: Cfg => Bench[SoftDut])(
+      tests: (Cfg, Bench[SoftDut]) => Unit
+  ) extends Nameable {
+
+    /** Override this function to define how configurations should appear in test results */
+    def cfgToString(cfg: Cfg): String = cfg.toString()
+
+    /** Run the tests with (a) component configuration(s) */
+    def runWith(cfgs: Cfg*): Unit = runWithAll(cfgs)
+
+    /** Run the tests with a sequnece of component configurations */
+    def runWithAll(cfgs: Seq[Cfg]): Unit = {
+      getName - {
+        for (cfg <- cfgs) {
+          cfgToString(cfg) - {
+            tests(cfg, benchBuilder(cfg))
+          }
+        }
+      }
+    }
+  }
+
+  /** Create a test-bench
+    *
+    * @param dut The device to test
```

Comment:
strictly speaking its not a component/device that is passed here ;-) not sure it should be mentioned/explained though...

But that's a fun thing, didn't know that scala would delay evaluation like this here... I was initially a bit stumped by the `Identity(IdentityConfig(0))` example below and why it didn't blow up immediately, so it might be worth highlighting that somewhat in the documentation later.

Reply:
IMHO and if I understand correctly, this distinction is a little bit advanced (for people beginning with SpinalHDL without a Scala background) Scala stuff from the `=> Dut` type (parameter by name); SpinalHDL makes things transparent for users so that they don't have to worry about this. Else it is pure Scala stuff so it does not have to be documented in SpinalHDL?

Maybe it is more about the delayed "lazy" compilation which makes the error pop when the user does not expect it? In this case maybe the error message is explicit enough? Idk… For instance in RTD it can be documented along with the added `it compiles` test, stating that the given expression is evaluated only during this test?

Reply:
@numero-744 yes, I only meant the delayed/lazy compilation, no need to mention the mechanics

Reply:
Okay so I have added an item for that in https://github.com/SpinalHDL/SpinalDoc-RTD/issues/155